execution/src/vm.rs

```diff
@@ -0,0 +1,298 @@
+                    } = operation.content.op
+                    {
+                        // get sender address
+                        // TODO remove unwrap
+                        let sender_addr =
+                            Address::from_public_key(&operation.content.sender_public_key).unwrap();
+
+                        // found an ExecuteSC operation
+                        let init_changes = {
+                            // get execution context guard
+                            let mut exec_context_guard =
+                                self.current_execution_context.0.lock().unwrap();
+
+                            // credit the sender with "coins"
+                            // TODO do not ignore result
+                            let _result = (*exec_context_guard).ledger_step.set_balance_delta(
+                                sender_addr,
+                                coins,
+                                true,
+                            );
+
+                            // credit the block creator with max_gas*gas_price
+                            // TODO consider dispatching with edorsers/endorsed as well
+                            // TODO remove unwrap()
+                            // TODO do not ignore result
+                            let _result = (*exec_context_guard).ledger_step.set_balance_delta(
+                                block_creator_addr,
+                                gas_price.checked_mul_u64(max_gas).unwrap(),
+                                true,
+                            );
+
+                            // save caused changes
+                            (*exec_context_guard).ledger_step.caused_changes.clone()
+
+                            // drop execution context guard
+                        };
+
+                        // parse operation bytecode
+                        match Module::from_binary(&self.store, &data) {
+                            Ok(module) => {
+                                (module, max_gas, coins, gas_price, init_changes, sender_addr)
+                            }
+                            Err(err) => {
+                                // parsing failed
+                                debug!(
+                                    "failed parsing bytecode in operation index {} in block {}: {}",
+                                    op_idx, block_id, err
+                                );
+                                continue;
+                            }
+                        }
+                    } else {
+                        // not an ExecuteSC operations
+                        continue;
+                    };
+
+                // a module was successfully parsed and balances are ready
+                // init_changes contains a copy of the initial coin credits that will be popped back if bytecode execution fails in order to cancel its effects
+
+                // fill context for execution
```

Comment:
I don't think this is necessary yet, since we don't even know how to pass the context properly to the executing code.

Reply:
it's mostly an example of how it could be done. Subject to change.